src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,790 @@
+
+
+cdef class fq_default(flint_scalar):
+    def __cinit__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+        fq_default_init(self.val, self.ctx.val)
+
+    def __dealloc__(self):
+        if self.ctx is not None:
+            fq_default_clear(self.val, self.ctx.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+
+        check = self.ctx.set_any_as_fq_default(self.val, val)
+        if check is NotImplemented:
+            raise TypeError
+
+    def __int__(self):
+        """
+        Attempts to lift self to an integer of type fmpz in [0, p-1]
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> int(gf(123))
+            123
+        """
+        cdef fmpz x = fmpz.__new__(fmpz)
+        res = fq_default_get_fmpz(x.val, self.val, self.ctx.val)
+        if res == 1:
+            return int(x)
+        raise ValueError("fq element has no lift to the integers")
+
+    def polynomial(self):
+        """
+        Returns a representative of ``self`` as a polynomial in `(Z/pZ)[x] / h(x)`
+        where `h(x)` is the defining polynomial of the finite field.
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> a.polynomial()
+            3*x^2 + 2*x + 1
+            >>> gf(123).polynomial()
+            123
+        """
+        cdef fmpz_mod_poly pol
+
+        ring_ctx = fmpz_mod_poly_ctx(self.ctx.prime())
+        pol = ring_ctx.new_ctype_poly()
+        fq_default_get_fmpz_mod_poly((<fmpz_mod_poly>pol).val, self.val, self.ctx.val)
+
+        return pol
+
+    def str(self):
+        return self.polynomial().str(var=self.ctx.var.decode())
+
+    def __repr__(self):
+        # TODO: what do we want here?
+        return str(self)
+
+    # =================================================
+    # Comparisons
+    # =================================================
+    def is_zero(self):
+        """
+        Returns true is self is zero and false otherwise
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> gf(0).is_zero()
+            True
+            >>> gf(-1).is_zero()
+            False
+        """
+        return 1 == fq_default_is_zero(self.val, self.ctx.val)
+
+    def is_one(self):
+        """
+        Returns true is self is one and false otherwise
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> gf(-1).is_one()
+            False
+            >>> gf(1).is_one()
+            True
+        """
+        return 1 == fq_default_is_one(self.val, self.ctx.val)
+
+    def __richcmp__(self, other, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fq_default cannot be ordered")
+
+        if not typecheck(other, fq_default):
+            other = self.ctx.any_as_fq_default(other)
+
+        if typecheck(other, fq_default):
```

Comment:
I'm not sure about using `any_as` in the `__richcmp__` context any more. I recently added something like this to `nmod_poly` and then afterwards thought that there must be a better way.

Correct me if I am wrong here but if  `F = fq(p^n)` then an element `x` of `F` is basically a polynomial over `G = fq(p)`. The only way that `x` is equal to anything that is not an element of `F` is if its polynomial is constant so that we can equate it to an element of `fq(p)` (or `Z/(p)`).

Is that correct?

If that is correct then I think it makes more sense rather than converting `other` to an `fq` we should check if self is a constant polynomial and if so downgrade `self` to an `fq(p)` or an `fmpz_mod(p)` or something for the comparison.

Or does it get more complicated than that? Are there conversions between e.g. `fq(p^d)` and `fq(p^e)`? Would there be a reason to equate elements of fields with same characteristic but different order?

We have to be careful here because `any_as_fq_default` is quite greedy but I think we want to be strict with `__eq__` in general. I'm not sure exactly what makes sense when it comes to equating finite field elements with other things that are not finite field elements or that are elements of different finite fields.

Does it make sense to seamlessly mix `nmod` and `fmpz_mod` in with `fq` types?

Also note this:
```python
In [8]: from flint import *

In [9]: two = nmod(2, 5)

In [10]: two == 2
Out[10]: True

In [11]: two == fmpz(2)
Out[11]: False

In [12]: fmpz(2) == fmpq(2)
Out[12]: True
```
I'm not entirely sure what is best here but I had the impression that this was an intentional decision that although `int` can mix in with nmod the flint types are clearly separated for zero vs nonzero characteristic. I don't know how that should extend to the case of mixing `nmod` etc with finite fields.

Reply:
yeah this is something we need to think about.

For GF(p) (when the degree is one) I think it's natural to allow:

fq_default(x) == fmpz_mod(x) == nmod(x) 

providing that the characteristic for fq_default matches the moduli for fmpz_mod, nmod -- in fact, even in FLINT the fmpz_mod type is used in some cases when the degree d = 1

For GF(p^d) for d != it's a little tricky there's two things which mathematically are interesting but im not sure how flint handles them.

1) If GF(p^d), GF(p^e) with e a divisor of d then there is an embedding of GF(p^e) into GF(p^d) and so elements of GF(p^e) all fit into GF(p^d). This embedding is only up to isomorphism though, because they could have different moduli

2) I suppose a special case of the above is when d = e but the moduli do not match -- there is a sense of equality here, but elements are only equal up to isomorphism, not just by comparing the polynomials.

I don't immediately see these fq_isomorphisms made available... but maybe they're here somewhere?

I think what might be sensible for now is:

1) if two elements are fq_default and have EXACTLY the same context then we can compare them
2) if one element is fq_default and another is fmpz_mod or nmod then they're equal if: a) the fq_default element has a lift to an integer AND the moduli match
3) if one element is fq_default and another is a int-like type, then they should match if int(fq_default) matches the int like type
4) if an fmpz_poly or nmod_poly has EXACTLY the same polynomial, then we could say these two things are equal? this is a little weird maybe?

Reply:
Of course another option is to just allow equality on the type fq_default and not allow any other coercion

Reply:
Oh! There's a whole thing for isomorphisms! This is what we should use for these checks then: https://flintlib.org/doc/fq_embed.html and not use equality here :)

Reply:
OK i think I agree with this then

I'll summarise what I think could work.

- `fq_default` elements are only equal if their value is equal and their contexts are equal. This means `fq_default(p).one() != fq_default(p, d).one()` as they're in different fields (assume d != 1)
- `fq_default` elements are equal to `fmpz_mod` and `nmod` elements if the degree of the finite field is one and the prime matches `fq_default(p).one() == nmod(1, p)`

A trickier one, which requires changing `fmpz` maybe would be:

- `fmpz(x) = int(x)` for all `x`

as an example of this:

```
import gmpy2
assert gmpy2.mpz(1) == 1
```

holds and I think we should match this.

If this is true, then we need to decide about whether:

```
fmpz(1) == nmod(1, p) == fmpz_mod(1, p) == fq_default(1, p)
fmpz(-11) == nmod(p-1, p) == fmpz_mod(p-1, p) == fq_default(p-1, p)
```

When the degree of `fq_default` is one

Reply:
This is what sage does:

```py
sage: R = Zmod(11)
sage: F = GF(11)
sage: R(1) == F(1) == int(1) == ZZ(1)
True
sage: R(-1) == F(-1) == int(-1) == ZZ(-1)
True
sage: R(10) == F(10) == int(10) == ZZ(10)
```

```py
sage: F = GF(11)
sage: F2 = GF(11**2)
sage: F(1) == F2(1)
True
sage: F(1) == F2(1) == ZZ(1) == int(1)
True
```

So they allow two elements to be coerced:

```py
sage: F2 = GF(11**2)
sage: F3 = GF(11**3)
sage: F2(1) == F3(1)
True
sage: F2(-1) == F3(-1)
True
```

Reply:
> At the moment we get something very very strange:

And also:
```python
In [9]: fmpz(2) == 2 == nmod(2, 3) != fmpz(2)
Out[9]: True

In [11]: fmpz(2) == 2 == 2.0 != fmpz(2)
Out[11]: True
```
If different pairs of objects use different notions of equality than transitivity is lost.

I doubt there is a unique good way to handle all of this especially when some things like `2 == 2.0` are outside of our control.

I think it is worth considering how other systems like Sage etc handle this sort of thing (not that any other system should be presumed "correct").

Reply:
I think `mpz` having `mpz(1) == 1` convinces me that `fmpz(x) == int(x)` should always hold

I think if I have an element `nmod(x, p)` or `fmpz_mod(x, p)` etc it's natural to think of all equality modulo `p` so if we compare `nmod` types to `fmpz` or `int` types then the integer should be reducxed modulo p before comparison which i think would mean our design aligns with the sage snippet above

Reply:
So maybe fq_defauly comparison could be something like...
```
if typecheck(other, fq_default):
    if element lifts to integers, only check value and prime matches
    else check value and whole context matches
elif typecheck(other fmpz):
    cast other to fq_default to reduce mod p and check equality
elif typecheck(other int):
    cast other to fq_default to reduce mod p and check equality
elif typecheck(other fmpz_mod):
    check modulus matches and value matches
elif typecheck(other nmod):
    check modulus matches and value matches
```

Reply:
> I think `mpz` having `mpz(1) == 1` convinces me that `fmpz(x) == int(x)` should always hold

So you would change the first two here:
```python
In [1]: from flint import *

In [2]: fmpz(2) == 2.0
Out[2]: False

In [3]: fmpz(2) == nmod(2, 3)
Out[3]: False

In [4]: fmpz(2) == int(2.0)
Out[4]: True

In [5]: fmpz(2) == int(nmod(2, 3))
Out[5]: True
```
Then we also have:
```python
In [6]: nmod(2, 3) == -1
Out[6]: True

In [7]: nmod(2, 3) == fmpz(-1)
Out[7]: False

In [8]: int(nmod(2, 3)) == fmpz(-1)
Out[8]: False
```
Perhaps it makes more sense to go the other way and say that when it comes to characteristic `p` the fmpz should be reduced modulo `p` for the comparison rather than lifting the nmod to fmpz. That is what happens for `nmod == int`.

Reply:
> Perhaps it makes more sense to go the other way and say that when it comes to characteristic `p`

If you think of `GF(p)` as `ZZ/(p)` then in some sense `ZZ` is the "ground domain" so it makes sense to promote from `ZZ` to `GF(p)` rather than vice versa...